src/symmetric.rs

```diff
@@ -471,6 +643,94 @@ impl SymmetricAlgorithmKey {
     }
 }
 
+#[cfg(feature = "block-cipher-trait")]
+pub use block_cipher;
+#[cfg(feature = "block-cipher-trait")]
+mod block_cipher_trait {
+    use super::*;
+
+    use core::convert::TryFrom;
+
+    use block_cipher::generic_array::{
+        typenum::{self, Unsigned},
+        ArrayLength,
+    };
+    use block_cipher::{self, Block, BlockCipher, Key, NewBlockCipher};
+
+    impl KeyBits for typenum::U128 {
+        const VALUE: Option<usize> = Some(Self::USIZE);
+    }
+    impl KeyBits for typenum::U192 {
+        const VALUE: Option<usize> = Some(Self::USIZE);
+    }
+    impl KeyBits for typenum::U256 {
+        const VALUE: Option<usize> = Some(Self::USIZE);
+    }
+
+    #[test]
+    fn assert_aes_keysize_valid() {
+        fn _inner<T: NewBlockCipher>() {}
+        _inner::<super::Key<Aes, typenum::U128>>();
+        _inner::<super::Key<Aes, typenum::U192>>();
+        _inner::<super::Key<Aes, typenum::U256>>();
+    }
+
+    impl<B: KeyBits> NewBlockCipher for super::Key<Aes, B>
+    where
+        B: typenum::Unsigned,
+        // Fancy way of allowing only {128, 192, 256}
+        B: typenum::IsGreaterOrEqual<typenum::U128, Output = typenum::B1>,
+        B: typenum::IsLessOrEqual<typenum::U256, Output = typenum::B1>,
+        B: typenum::PartialDiv<typenum::U64>,
+        // Help the trait solver see that it's also divisible by 8
+        B: typenum::PartialDiv<typenum::U8>,
+        <B as typenum::PartialDiv<typenum::U8>>::Output: ArrayLength<u8>,
+    {
+        /// Key size in bytes with which cipher guaranteed to be initialized.
+        type KeySize = <B as typenum::PartialDiv<typenum::U8>>::Output;
+
+        /// Create new block cipher instance from key with fixed size.
+        fn new(key: &Key<Self>) -> Self {
+            let prov =
+                SymmetricAlgorithm::open(SymmetricAlgorithmId::Aes, ChainingMode::Ecb).unwrap();
+            let key = prov.new_key(key).unwrap();
+            match Self::try_from(key) {
+                Ok(value) => value,
+                Err(..) => panic!(),
+            }
+        }
+    }
+
+    impl<B: KeyBits> BlockCipher for super::Key<Aes, B> {
+        /// Size of the block in bytes
+        type BlockSize = typenum::U16;
+
+        /// Number of blocks which can be processed in parallel by
+        /// cipher implementation
+        type ParBlocks = typenum::U1;
+
+        /// Encrypt block in-place
+        fn encrypt_block(&self, block: &mut Block<Self>) {
+            // FIXME: We assume here that we use ECB (as initialized via the
+            // NewBlockCipher trait)
+            // FIXME: Adapt the implementation to use the in-place one
```

Comment:
Indeed, that would be nice to have. An assert on the size of the provided array will be required.

Reply:
Do you want to block this PR on having this implemented using a new in-place API? I'd prefer to implement that as an improvement in a separate PR, if you don't mind

Reply:
Yeah, let's do it in another PR. I created #32 to not forget about it.